cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/CachingPodsObserver.java

```diff
@@ -0,0 +1,132 @@
+ */
+
+package com.netflix.spinnaker.cats.redis.cluster;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.netflix.spinnaker.cats.agent.Agent;
+import com.netflix.spinnaker.cats.cluster.NodeIdentity;
+import com.netflix.spinnaker.cats.cluster.ShardingFilter;
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService;
+import com.netflix.spinnaker.kork.jedis.RedisClientDelegate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachingPodsObserver implements ShardingFilter, Runnable {
+
+  private static final Logger logger = LoggerFactory.getLogger(CachingPodsObserver.class);
+  private int podCount = 0;
+  private int podIndex = -1;
+  private static final String REPLICA_SSET_KEY = "clouddriver:caching:replicas";
+  private static final String CORE_PROVIDER =
+      "com.netflix.spinnaker.clouddriver.core.provider.CoreProvider";
+  private final RedisClientDelegate redisClientDelegate;
+  private final NodeIdentity nodeIdentity;
+  private final long replicaKeyTtl;
+  // this script adds or updates a unique id as a member of a sorted set with score equal to current
+  // time plus
+  // sharding.replica-key-ttl-seconds, deletes the members having scores less than current time(ms)
+  // and finally
+  // fetches list of all members of the sorted set which represent the live caching pods
+  private static final String HEARTBEAT_REFRESH_SCRIPT =
+      "redis.call('zadd', KEYS[1], ARGV[1], ARGV[2])"
+          + " redis.call('zremrangebyscore', KEYS[1], '-inf', ARGV[3])"
+          + " return redis.call('zrange', KEYS[1], '0', '-1')";
+
+  public CachingPodsObserver(
+      RedisClientDelegate redisClientDelegate,
+      NodeIdentity nodeIdentity,
+      DynamicConfigService dynamicConfigService) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.nodeIdentity = nodeIdentity;
+    long observerIntervalSeconds =
+        dynamicConfigService.getConfig(Integer.class, "sharding.observer-interval-seconds", 30);
+    replicaKeyTtl =
+        dynamicConfigService.getConfig(Integer.class, "sharding.replica-key-ttl-seconds", 60);
+    ScheduledExecutorService podsObserverExecutorService =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat(CachingPodsObserver.class.getSimpleName() + "-%d")
+                .build());
+    podsObserverExecutorService.scheduleAtFixedRate(
+        this, 0, observerIntervalSeconds, TimeUnit.SECONDS);
+    refreshHeartbeat();
+    logger.info("Account based sharding is enabled for all caching pods.");
+  }
+
+  @Override
+  public void run() {
+    try {
+      refreshHeartbeat();
+    } catch (Exception e) {
```

Comment:
Catch `Throwable` like in sql implementation. This is because the ExecutorService will stop working if there's ever an unhandled exception.

Reply:
Will do.

Reply:
done